### PR TITLE
Add opCodeForCorresponding APIs for load/store

### PR DESCRIPTION
**Pull request by liqunl**
The APIs are needed for a PRE fix.

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
Depends on https://github.com/eclipse/openj9/pull/10358
@andrewcraik Could you review?

File: compiler/il/OMRIL.cpp [hunk 1]
```diff
@@ -652,6 +653,102 @@ OMR::IL::opCodeForCorrespondingIndirectStore(TR::ILOpCodes storeOpCode)
...
+      case TR::bload: return TR::bstore;
+      case TR::sload: return TR::sstore;
+      case TR::iload: return TR::istore;
+      case TR::lload: return TR::lstore;
+      case TR::fload: return TR::fstore;
+      case TR::dload: return TR::dstore;
+      case TR::aload: return TR::astore;
+      case TR::vload: return TR::vstore;
+      case TR::cload: return TR::cstore;
+      case TR::buload: return TR::bstore;
```
Comment (andrewcraik) [inline compiler/il/OMRIL.cpp:694]: these unsigned cases are deprecated and no longer generated I believe... @0xdaryl that's right isn't it?

File: compiler/il/OMRIL.cpp [hunk 2]
```diff
@@ -652,6 +653,102 @@ OMR::IL::opCodeForCorrespondingIndirectStore(TR::ILOpCodes storeOpCode)
...
+      case TR::srdbar:
+      case TR::irdbar:
+      case TR::lrdbar:
+      case TR::frdbar:
+      case TR::drdbar:
+      case TR::ardbar:
+         //There is not necessarily a guaranteed symmetry about whether an direct rdbar should be mapped to
+         //an direct wrtbar or a normal direct store. The mapping of rdbar/ wrtbar totally depends on the
+         //actual use in subprojects and should be undefined in OMR level.
+         TR_ASSERT_FATAL(0, "xrdbari can't be used with global opcode mapping API at OMR level\n");
```
Comment (andrewcraik) [inline compiler/il/OMRIL.cpp:707]: Assert message references the wrong family of opcodes.

File: compiler/il/OMRIL.cpp [hunk 3]
```diff
@@ -652,6 +653,102 @@ OMR::IL::opCodeForCorrespondingIndirectStore(TR::ILOpCodes storeOpCode)
...
+      default: break;
+      }
+
+   TR_ASSERT(0, "no corresponding store opcode for specified load opcode");
```
Comment (andrewcraik) [inline compiler/il/OMRIL.cpp:711]: Shouldn't this be ASSERT_FATAL - we can't be doing anything sensible right?

File: compiler/il/OMRIL.cpp [hunk 4]
```diff
@@ -652,6 +653,102 @@ OMR::IL::opCodeForCorrespondingIndirectStore(TR::ILOpCodes storeOpCode)
...
+      case TR::bwrtbar:
+      case TR::swrtbar:
+      case TR::iwrtbar:
+      case TR::lwrtbar:
+      case TR::fwrtbar:
+      case TR::dwrtbar:
+         //There is not necessarily a guaranteed symmetry about whether an direct wrtbar should be mapped to
+         //an direct rdbar or a normal direct load. The mapping of rdbar/wrtbar totally depends on the
+         //actual use in subprojects and should be undefined in OMR level.
+         TR_ASSERT_FATAL(0, "xwrtbari can't be used with global opcode mapping API at OMR level\n");
```
Comment (andrewcraik) [inline compiler/il/OMRIL.cpp:743]: Again wrong family of opcode referenced

File: compiler/il/OMRIL.cpp [hunk 5]
```diff
@@ -652,6 +653,102 @@ OMR::IL::opCodeForCorrespondingIndirectStore(TR::ILOpCodes storeOpCode)
...
+      case TR::istore:  return TR::iload;
+      case TR::lstore:  return TR::lload;
+      case TR::fstore:  return TR::fload;
+      case TR::dstore:  return TR::dload;
+      case TR::astore:  return TR::aload;
+      case TR::awrtbar: return TR::aload;
+      case TR::vstore:  return TR::vload;
+      case TR::cstore:  return TR::sload;
+      case TR::bustore: return TR::bload;
+      case TR::iustore: return TR::iload;
```
Comment (andrewcraik) [inline compiler/il/OMRIL.cpp:732]: the unsigned stores are also deprecated and not generated - again @0xdaryl correct me if I am wrong please

File: compiler/il/OMRIL.cpp [hunk 6]
```diff
@@ -652,6 +653,102 @@ OMR::IL::opCodeForCorrespondingIndirectStore(TR::ILOpCodes storeOpCode)
...
+      case TR::dwrtbar:
+         //There is not necessarily a guaranteed symmetry about whether an direct wrtbar should be mapped to
+         //an direct rdbar or a normal direct load. The mapping of rdbar/wrtbar totally depends on the
+         //actual use in subprojects and should be undefined in OMR level.
+         TR_ASSERT_FATAL(0, "xwrtbari can't be used with global opcode mapping API at OMR level\n");
+
+      default: break;
+      }
+
+   TR_ASSERT(0, "no corresponding load opcode for specified store opcode");
```
Comment (andrewcraik) [inline compiler/il/OMRIL.cpp:748]: again shouldn't this be assert fatal?

**Comment by andrewcraik** [comment]
FYI @mstoodle and @0xdaryl since this is a new node/opcode API which needs appropriate review and an opinion other than my own. I think the API makes sense - it hides a lot of the complexity of converting loads and stores and can stop a proliferation of logic through the compiler testing for particular load/store types across the compiler before converting the opcode.

@liqunl could we split this into a change that adds the API with no uses and then doing a second PR which adds API uses so downstream projects don't have to do a coordinated merge of this change with updates to their own code to support this new API?

**Comment by liqunl** [comment]
Split this PR into two. The PR to use the APIs is #5461

File: compiler/il/OMRIL.cpp (hunk 2 above)
Comment (liqunl) [reply compiler/il/OMRIL.cpp]: Fixed

File: compiler/il/OMRIL.cpp (hunk 4 above)
Comment (liqunl) [reply compiler/il/OMRIL.cpp]: Fixed

File: compiler/il/OMRIL.cpp (hunk 3 above)
Comment (liqunl) [reply compiler/il/OMRIL.cpp]: Fixed.

File: compiler/il/OMRIL.cpp (hunk 6 above)
Comment (liqunl) [reply compiler/il/OMRIL.cpp]: Fixed

**Comment by andrewcraik** [comment]
@mstoodle / @vijaysun-omr do you wish to review/comment given the core API change here? I think it looks good and will launch a build pending any comments.

**Comment by andrewcraik** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
@andrewcraik asked me to review this and it looks reasonable to me.

**Pull request merged**